zebra/zebra_evpn_arp_nd.c

```diff
@@ -0,0 +1,651 @@
+static void zebra_evpn_arp_nd_pkt_dump(struct zebra_if *zif, uint16_t vlan,
+				       uint8_t *data, int len)
+{
+	struct ethhdr *ethh = (struct ethhdr *)data;
+
+	if (IS_ZEBRA_DEBUG_EVPN_MH_ARP_ND_PKT)
+		zlog_debug(
+			"evpn arp_nd pkt on %s vlan %d [dm=%pEA sm=%pEA et=0x%x]",
+			zif->ifp->name, vlan ? vlan : zif->pvid,
+			(struct ethaddr *)&ethh->h_dest,
+			(struct ethaddr *)&ethh->h_source,
+			ntohs(ethh->h_proto));
+	/* XXX - dump ARP/NA info */
+}
+
+void zebra_evpn_arp_nd_print_summary(struct vty *vty, bool uj)
+{
+	json_object *json = NULL;
+
+	if (uj) {
+		json = json_object_new_object();
+		json_object_boolean_true_add(json, "arpRedirect");
+		json_object_int_add(json, "arpReplyPkts",
+				    zevpn_arp_nd_info.stat.arp);
+		json_object_int_add(json, "naPkts", zevpn_arp_nd_info.stat.na);
+		json_object_int_add(json, "redirectPkts",
+				    zevpn_arp_nd_info.stat.redirect);
+		json_object_int_add(json, "notReadyPkts",
+				    zevpn_arp_nd_info.stat.not_ready);
+		json_object_int_add(json, "vniMissingPkts",
+				    zevpn_arp_nd_info.stat.vni_missing);
+		json_object_int_add(json, "macMissingPkts",
+				    zevpn_arp_nd_info.stat.mac_missing);
+		json_object_int_add(json, "esNonLocalPkts",
+				    zevpn_arp_nd_info.stat.es_non_local);
+		json_object_int_add(json, "esUpPkts",
+				    zevpn_arp_nd_info.stat.es_up);
+	} else {
+		vty_out(vty, "EVPN ARP-reply/NA redirect: %s\n",
+			(zevpn_arp_nd_info.flags & ZEBRA_EVPN_ARP_ND_FAILOVER)
+				? "enabled"
+				: "disabled");
+		vty_out(vty, "Stats:\n");
+		vty_out(vty, "  IPv4 ARP replies: %u\n",
+			zevpn_arp_nd_info.stat.arp);
+		vty_out(vty, "  IPv6 neighbor advertisements: %u\n",
+			zevpn_arp_nd_info.stat.na);
+		vty_out(vty, "  Redirected packets: %u\n",
+			zevpn_arp_nd_info.stat.redirect);
+		vty_out(vty, "  Skipped packets:\n");
+		vty_out(vty, "    Not ready: %u\n",
+			zevpn_arp_nd_info.stat.not_ready);
+		vty_out(vty, "    VNI missing: %u\n",
+			zevpn_arp_nd_info.stat.vni_missing);
+		vty_out(vty, "    MAC missing: %u\n",
+			zevpn_arp_nd_info.stat.mac_missing);
+		vty_out(vty, "    Dest is not local ES: %u\n",
+			zevpn_arp_nd_info.stat.es_non_local);
+		vty_out(vty, "    Dest ES oper-up: %u\n",
+			zevpn_arp_nd_info.stat.es_up);
+	}
+
+	if (uj) {
+		vty_out(vty, "%s\n",
+			json_object_to_json_string_ext(
+				json, JSON_C_TO_STRING_PRETTY));
+		json_object_free(json);
+	}
+}
+
+void zebra_evpn_arp_nd_if_print(struct vty *vty, struct zebra_if *zif)
+{
+	if (zif->arp_nd_info.pkt_fd > 0)
+		vty_out(vty,
+			"  ARP-ND redirect enabled: ARP-replies %u NA %u\n",
+			zif->arp_nd_info.arp_pkts, zif->arp_nd_info.na_pkts);
+}
+
+/* Send to the ES peer VTEP-IP */
+static void zebra_evpn_arp_nd_udp_send(struct in_addr vtep_ip, uint8_t *data,
+				       int len)
+{
+	struct sockaddr_in sin;
+
+	memset(&sin, 0, sizeof(sin));
+	sin.sin_family = AF_INET;
+	/* XXX - the VxLAN UDP port is user configurable so we
```

Comment:
Is this coming in a  follow-up?

Reply:
Sure, I will open a ticket to track that.